src/components/modals/ActionPayloadEditor/slateTransformer.ts

```diff
@@ -46,14 +47,108 @@ function updateNodeOptionNames(node: any, options: IOption[], newName?: string)
         if (typeof newName === 'string') {
             node.leaves.forEach((n: any) => n.text = `$${newName}`)
         }
-        
+
         return node
     } else {
         node.nodes.map((n: any) => updateNodeOptionNames(n, options))
         return node
     }
 }
 
+/**
+ * Replaces entity nodes with values. The values name be entity names, memory values, etc.
+ * This is more flexible than the other functions which only replace with fixed thing name or memory and cant recognize missing entities
+ * Hopefully the other functions will be retired and replaced in future.
+ *
+ * @param node Slate node (Value, Document, Block, Inline, Text, etc)
+ * @param entityValuesMap Map from entityId to entityEntry which has name and ?value
+ * @param replacerFn Function to be run on each replacement of entity node. Can be anything, but usually replaces with entity name or memory value
+ * @param entitiesRequired Boolean indicating if `entityMissing` flag should be added when entity without memory value is present
+ */
+function replaceEntityNodesWithValues(node: any, entityValuesMap: Record<string, Util.EntityMapEntry>, replacerFn: (entityEntry: Util.EntityMapEntry) => string, entitiesRequired: boolean): any {
```

Comment:
I think it wold be more appropriate to do this operation in `Slate` domain instead of JSON since they have their own API and normalizers than ensure allowed operations are done, but I ended up doing in JSON domain.

Might need to revisit this in future.